### PR TITLE
Using PKCE flow for Gemini Authentication, adding shared oauth utility

### PR DESCRIPTION
**Pull request by ryanml**
Fixes https://github.com/brave/brave-browser/issues/11043

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
General authentication regression for Binance and Gemini

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: browser/BUILD.gn [hunk 1]
```diff
@@ -312,6 +312,12 @@ source_set("browser_process") {
     ]
   }
 
+  if (gemini_enabled || binance_enabled) {
+    deps += [
+      "//brave/components/crypto_exchange/browser",
```
Comment (ryanml) [inline browser/BUILD.gn:317]: We can use this module for more code sharing utilities

File: components/gemini/browser/gemini_service_browsertest.cc [hunk 2]
```diff
@@ -313,12 +313,18 @@ IN_PROC_BROWSER_TEST_F(GeminiAPIBrowserTest, GetOAuthClientURL) {
     "response_type=code&"
     "client_id=fake-client-id&"
     "redirect_uri=com.brave.gemini%3A%2F%2Fauthorization&"
-    "scope=trader&"
+    "scope=addresses%3Aread%2Cbalances%3Aread%2Corders%3Acreate&"
```
Comment (ryanml) [inline components/gemini/browser/gemini_service_browsertest.cc:316]: I neglected to update this with the recent scope change

File: components/crypto_exchange/browser/crypto_exchange_oauth_util.cc [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  std::string code_challenge;
+  char raw[crypto::kSHA256Length] = {0};
+  crypto::SHA256HashString(code_verifier,
+                           raw,
+                           crypto::kSHA256Length);
+  base::Base64Encode(base::StringPiece(raw,
+                                       crypto::kSHA256Length),
+                                       &code_challenge);
+
+  if (strip_chars) {
```
Comment (ryanml) [inline components/crypto_exchange/browser/crypto_exchange_oauth_util.cc:44]: This works for Binance, but Gemini expects characters preserved so these steps are optional

File: browser/BUILD.gn [hunk 4]
```diff
@@ -312,6 +312,12 @@ source_set("browser_process") {
     ]
   }
 
+  if (gemini_enabled || binance_enabled) {
```
Comment (ryanml) [inline browser/BUILD.gn:315]: Actually this should probably be a gn var like `crypto_exchanges_enabled`

Comment (ryanml) [reply browser/BUILD.gn:315]: I'll update after the rest of the review

Comment (ryanml) [reply browser/BUILD.gn]: updated

**Comment by kjozwiak** [comment]
Verification **PASSED** on `macOS 10.15.6 x64` using the following build:

```
Brave | 1.14.44 Chromium: 85.0.4183.69 (Official Build) nightly (64-bit)
-- | --
Revision | 4554ea1a1171bd8d06951a4b7d9336afe6c59967-refs/branch-heads/4183@{#1426}
OS | macOS Version 10.15.6 (Build 19G73)
```

* Went through a general regression pass for Gemini authentication via https://github.com/brave/brave-core/pull/6280#issuecomment-674324364

Went through the following cases via the `Binance` widget as per https://github.com/brave/brave-core/pull/6306#issue-460583728:

* ensured that you can connect without any issues and the `Summary`, `Deposit`, `Convert` and `Buy` tabs are displayed
* ensured that disconnecting/reconnecting via the widget worked as expected and displayed the correct blanaces
* ensured that deposit addresses are being displayed correctly under the `Deposit` tab
* ensured that clicking on the QR code images under the `Deposit` tab correctly enlarges/displays the QR codes
* ensured that tokens such as HBAR display both a deposit address and memo address
* ensured that selecting `Learn More` under the kabob/settings menu opened https://brave.com/binance
* ensured that selecting tokens via the `Buy` tab correctly updates the `Buy [Token]` button
* ensured that clicking on the `Buy` button under the `Buy` tab correctly opens a new tab with the correct amounts already filled out

**Pull request merged**